### PR TITLE
[REJECTED] Restore Stream-like lazy tail construction [ci: last-only]

### LINKED ISSUE #11984: LazyList.isEmpty forces tail
I am trying to replace usages of the deprecated Stream with the recommended LazyList. But my code is now failing because it turns out that calling `isEmpty` on a LazyList actually forces the tail.

## reproduction steps

Compile and run this code with Scala 2.13.2:

```scala
object StreamTest {

  def main(args: Array[String]): Unit = {
    println(stream2.isEmpty)
    println(lazyList2.isEmpty)
  }

  def stream1: Stream[Int] = {
    println("stream1")
    Stream(1)
  }
  def stream2: Stream[Int] = 2 #:: stream1

  def lazyList1: LazyList[Int] = {
    println("lazyList1")
    LazyList(1)
  }
  def lazyList2: LazyList[Int] = 2 #:: lazyList1
}
```

## problem

The output is the following, which shows that the tail argument of the LazyList is called, which is not the case for Stream.

```
false
lazylist1
false
```

## expectation

I expect the behavior of LazyList to be at least as lazy as Stream.

### PR DESCRIPTION
**Pull request by som-snytt**
The most important part is getting rid of the extraneous braces.

Fixes scala/bug#11984

### CODE REVIEW

File: src/library/scala/collection/immutable/LazyList.scala [hunk 1]
```diff
@@ -216,22 +216,19 @@ final class LazyList[+A] private(private[this] var lazyState: () => LazyList.Sta
     with Serializable {
   import LazyList._
 
-  @volatile private[this] var stateEvaluated: Boolean = false
```
Comment (som-snytt) [inline src/library/scala/collection/immutable/LazyList.scala:219]: lost the volatile on lazyState

**Comment by som-snytt** [comment]
Mima says, For your consideration:
```
[error]   - library/mimaReportBinaryIssues failed: java.lang.RuntimeException: Failed binary compatibility check against org.scala-lang:scala-library:2.13.0! Found 3 potential problems (filtered 432)
[error] Saved log of library/mimaReportBinaryIssues:
[error] scala-library: Failed binary compatibility check against org.scala-lang:scala-library:2.13.0! Found 3 potential problems (filtered 432)
[error]  * method this(java.lang.Object,scala.collection.immutable.LazyList)Unit in class scala.collection.immutable.LazyList#State#Cons's type is different in current version, where it is (java.lang.Object,scala.Function0)Unit instead of (java.lang.Object,scala.collection.immutable.LazyList)Unit
[error]    filter with: ProblemFilters.exclude[IncompatibleMethTypeProblem]("scala.collection.immutable.LazyList#State#Cons.this")
[error]  * method tailDefined()Boolean in class scala.collection.immutable.LazyList#State#Cons does not have a correspondent in other version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.LazyList#State#Cons.tailDefined")
[error]  * method this(java.lang.Object,scala.Function0)Unit in class scala.collection.immutable.LazyList#State#Cons's type is different in other version, where it is (java.lang.Object,scala.collection.immutable.LazyList)Unit instead of (java.lang.Object,scala.Function0)Unit
[error]    filter with: ProblemFilters.exclude[IncompatibleMethTypeProblem]("scala.collection.immutable.LazyList#State#Cons.this")
[error] testAll failed due to previous errors
[error] (testAll) testAll failed due to previous errors
```

**Comment by sjrd** [comment]
Please don't. This is adding an extra layer of indirection for no reason. The tail is *already* lazy by virtue of `LazyList` being a completely lazy thunk itself.

I definitely agree that we need an easier way to build such a lazy thunk in user space, but not too make the tail twice lazy instead.

**Comment by SethTisue** [comment]
@som-snytt is there a use case where this is preferable to changing `#::` _only_?

**Comment by som-snytt** [comment]
The direction I was going in was: let lazyState transition to first realizing head and then realizing tail.

The comments about "already lazy" correctly describe that currently newLL gets a thunk for State to compute both head and tail.

The combinators doing `newLL(sCons(myhead, mytail))` would need the smarter lazyState.

@SethTisue this is the change to `#::`, but just the minimal change to preserve the original thunkiness passed in by the user.

As nth & seb pointed out, the different approach is just to provide conveniences for deferred construction.

There are opportunities for surprise, such as `42 #:: LazyList.empty` isn't trivial. I think the doc even says, Maybe you shouldn't be using LazyList.

I won't pursue this branch. I think I started wondering about it because of the question from last week about where did these lambdas come from.

**Comment by som-snytt** [comment]
Also, obviously "draft mode" means nothing.

**Pull request closed without merging**